pandas/core/reshape.py

```diff
@@ -558,87 +549,138 @@ def _convert_level_number(level_num, columns):
 
     # this makes life much simpler
-    if level_num != frame.columns.nlevels - 1:
-        # roll levels to put selected level at end
-        roll_columns = this.columns
-        for i in range(level_num, frame.columns.nlevels - 1):
+    # roll levels to put selected level(s) at end
+    roll_columns = this.columns
+    for j, level_num in enumerate(reversed(level_nums)):
+        for i in range(level_num, frame.columns.nlevels - (j + 1)):
             # Need to check if the ints conflict with level names
             lev1 = _convert_level_number(i, roll_columns)
             lev2 = _convert_level_number(i + 1, roll_columns)
             roll_columns = roll_columns.swaplevel(lev1, lev2)
-        this.columns = roll_columns
+    this.columns = roll_columns
 
     if not this.columns.is_lexsorted():
         # Workaround the edge case where 0 is one of the column names,
-        # which interferes with trying to sort based on the first
-        # level
+        # which interferes with trying to sort based on the first level
         level_to_sort = _convert_level_number(0, this.columns)
         this = this.sortlevel(level_to_sort, axis=1)
 
-    # tuple list excluding level for grouping columns
-    if len(frame.columns.levels) > 2:
+    num_levels_to_stack = len(level_nums)
+    level_vals = this.columns.levels[-num_levels_to_stack:]
+    level_labels = sorted(set(zip(*this.columns.labels[-num_levels_to_stack:])))
+    level_vals_used = MultiIndex.from_tuples([tuple(np.nan if lab == -1 else level_vals[i][lab]
+                                                    for i, lab in enumerate(label))
+                                              for label in level_labels],
+                                             names=this.columns.names[-num_levels_to_stack:])
+    levsize = len(level_labels)
+
+    # construct new_index
+    N = len(this)
+    if isinstance(this.index, MultiIndex):
+        new_levels = list(this.index.levels)
+        new_names = list(this.index.names)
+        new_labels = [lab.repeat(levsize) for lab in this.index.labels]
+    else:
+        new_levels = [this.index]
+        new_labels = [np.arange(N).repeat(levsize)]
+        new_names = [this.index.name]  # something better?
+    new_levels += level_vals
+    new_labels += [np.tile(labels, N) for labels in zip(*level_labels)]
+    new_names += level_vals_used.names
+    new_index = MultiIndex(levels=new_levels, labels=new_labels,
+                           names=new_names, verify_integrity=False)
+
+    # if stacking all levels in columns, result will be a Series
+    if len(frame.columns.levels) == num_levels_to_stack:
+        new_data = frame.values.ravel()
+        if dropna:
+            mask = notnull(new_data)
+            new_data = new_data[mask]
+            new_index = new_index[mask]
+        return Series(new_data, index=new_index)
+
+    # result will be a DataFrame
+
+    def _equal_key(keya, keyb):
+        return all(((a == b) or
```

Comment:
Good suggestion, but I get the following. Not sure yet what the type is that it's complaining about...
```
======================================================================
ERROR: test_stack_nan_index (pandas.tests.test_frame.TestDataFrame)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\tests\test_frame.py", line 12872, in test_stack_nan_index
    result_nan = df_nan.stack(level, dropna=dropna, sequentially=sequentially)
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\core\frame.py", line 3443, in stack
    return stack_levels_sequentially(self, level_num, dropna=dropna)
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\core\reshape.py", line 504, in stack_levels_sequentially
    return stack_single_level(frame, level_nums, dropna=dropna)
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\core\reshape.py", line 472, in stack_single_level
    return stack_multi_levels_simultaneously(frame, level_nums=[level_num], dropna=dropna)
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\core\reshape.py", line 643, in stack_multi_levels_simultaneously
    for cv in columns_values])
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\core\reshape.py", line 643, in <listcomp>
    for cv in columns_values])
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\core\common.py", line 442, in array_equivalent
    return ((left == right) | (np.isnan(left) & np.isnan(right))).all()
TypeError: Not implemented for this type

======================================================================
ERROR: test_unstack_nan_index (pandas.tests.test_frame.TestDataFrame)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\tests\test_frame.py", line 12384, in test_unstack_nan_index
    left  = df.set_index(['jim', 'joe']).unstack()['jolie']
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\core\frame.py", line 3507, in unstack
    result = self.T.stack(level, dropna=False, sequentially=sequentially).T
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\core\frame.py", line 3441, in stack
    return stack_multi_levels_simultaneously(self, level_num, dropna=dropna)
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\core\reshape.py", line 643, in stack_multi_levels_simultaneously
    for cv in columns_values])
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\core\reshape.py", line 643, in <listcomp>
    for cv in columns_values])
  File "c:\Users\seth\github\pandas\build\lib.win-amd64-3.4\pandas\core\common.py", line 442, in array_equivalent
    return ((left == right) | (np.isnan(left) & np.isnan(right))).all()
TypeError: Not implemented for this type

----------------------------------------------------------------------
```